Make the `getMessage` function for the javascript example more robust.

qdirks opened the pull request:
The old getMessage function had a bug where if you made multiple unawaited calls to the function, it wouldn't behave correctly because previous reads had not yet completed, and all you were doing was reading 4 bytes at a time from the input into the header buffer. It also had a bug of not reading the correct amount of data, because the buffer given to the read function was not exactly sized for the expected message size, so it was possible that the amount of data that you got back was larger than the size that the header indicated if the extension had sent more than one message.

The new getMessage function solves both of those issues. The first issue is solved by using a promise and a while loop to trap the execution of extra calls at the top of the function while there is already a read in progress. Each call to getMessage is let through the trap one at a time as a read completes. If there are no pending reads, then a call to getMessage does not get trapped at the top of the function, because the read promise will be null, and so the loop will be skipped. The other issue is solved by using buffers which are appropriately sized according to the header information. I think this is also more succinct, since the old `readFullAsync` ultimately ended up doing that anyway at the end by passing the data array to the Uint8Array constructor. If you just create it from the header information and pass it directly to the read function, you can avoid dispatching to another function, and it prevents a bug.

<!-- 🙌 Thanks for contributing to MDN Web Docs. Adding details below will help us to merge your PR faster. -->

### Description

<!-- ✍️ Summarize your changes in one or two sentences -->
Make the getMessage function better for the javascript example.

### Motivation

<!-- ❓ Why are you making these changes and how do they help readers? -->
The change will prevent the bugs I described, which makes it easier to develop a native messaging extension for the next person, I hope.

### Additional details

<!-- 🔗 Link to release notes, vendor docs, bug trackers, source control, or other places providing more context -->
https://nodejs.org/api/fs.html#filehandlereadbuffer-options

### Related issues and pull requests

<!-- 🔨 If this fully resolves a GitHub issue, use "Fixes #123" -->
<!-- 👉 Highlight related pull requests using "Relates to #123" -->
<!-- ❗ If another pull request should be merged first, use "**Depends on:** #123" -->


<!-- 👷‍♀️ After submitting, go to the "Checks" tab of your PR for the build status -->

qdirks committed:
Make the `getMessage` function for the javascript example more robust.

The old getMessage function had an awful bug where if you made multiple unawaited calls to the function, it wouldn't behave correctly because previous reads had not yet completed, and all you were doing was reading 4 bytes at a time from the input into the header buffer. It also had an awful bug of not reading the correct amount of data, because the buffer given to the read function was not exactly sized for the expected message size, so it was possible that the amount of data that you got back was larger than the size that the header indicated if the extension had sent more than one message.

The new getMessage function solves both of those issues. The first issue is solved by using a promise and a while loop to trap the execution of further calls at the top of the function while there is already a read in progress. Each call to getMessage is let through the trap one at a time as a read completes. If there are no pending reads, then a call to getMessage does not get trapped at the top of the function, because the read promise will be null, and so the loop will be skipped. The other issue is solved by using buffers which are appropriately sized according to the header information. I think this is also more succinct, since the old `readFullAsync` ultimately ended up doing that anyway at the end by passing the data array to the Uint8Array constructor. If you just create it from the header information and pass it directly to the read function, you can avoid dispatching to another function, and it prevents a bug.

qdirks committed:
Merge branch 'main' into qdirks-patch-1

qdirks commented on files/en-us/mozilla/add-ons/webextensions/native_messaging/index.md:
```diff
@@ -244,27 +244,20 @@ You can quickly get started sending and receiving messages with this NodeJS code
 
 import fs from "node:fs/promises";
 
+const _r = rs=>_rs_rp = rs; let _rs_rp, _rp = null; // read promise
```
I prefer the one liner. If you don't like it, then change it to whatever it needs to be. I don't know how to do that.

qdirks commented on files/en-us/mozilla/add-ons/webextensions/native_messaging/index.md:
```diff
@@ -244,27 +244,20 @@ You can quickly get started sending and receiving messages with this NodeJS code
+  const message = new Uint8Array(header[0]);
+  await input.read(message);
+  await input.close();
+  _rs_rp(); _rp = null; // allow the next getMessage call to proceed
```
I prefer the one liner. I resolve the promise, then I set it to null. In my head, it makes sense to be on one line, but if you don't like it then change it. Thank you.

Josh-Cena reviewed (dismissed):
Please make sure your code follows our [style guide](https://developer.mozilla.org/en-US/docs/MDN/Writing_guidelines/Code_style_guide/JavaScript). In particular:

1. Do not use underscore-prefixed variables, and use camelCase.
2. Names like `rs_rp` and `rp` don't mean anything to me. Use full words that make sense.
3. There's no bug about "unawaited calls", because all calls are awaited in our example (we show you how `getMessage` should be used). If your calls to `getMessage` are different then that's really a bug in your code. You seem to be implementing a lock with your `_rp` but again the variables are poorly named and I can't read the code. I wouldn't recommend implementing a lock yourself though; it's too much mechanism for a trivial demo.

I understand the part about the size mismatch though. I don't know why the original version is using a loop instead of preallocating the whole thing.

dotproto committed:
Apply linter suggestions from code review

Changes made by [mdn-linter] in PR mdn/content#39672.

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

qdirks committed:
Merge branch 'main' into qdirks-patch-1

qdirks committed:
Make recommended changes provided from feedback

Change the naming of variables to match the style on MDN. Try to make the code a little easier to understand.

qdirks committed:
Merge branch 'main' into qdirks-patch-1

Josh-Cena committed:
Update index.md

qdirks committed:
Merge branch 'main' into qdirks-patch-1

rebloor commented on files/en-us/mozilla/add-ons/webextensions/native_messaging/index.md:
```diff
@@ -244,25 +244,29 @@ You can quickly get started sending and receiving messages with this NodeJS code
-async function readFullAsync(length, buffer = new Uint8Array(65536)) {
-  const data = [];
-  while (data.length < length) {
+// This function cannot be called multiple times in parallel because they
+// share access to /dev/stdin. We use a lock `readPromise` to prevent this.
```
Perhaps I'm not fully understanding this comment, but it seems as though it should read like this:
```suggestion
// This function should not be called in parallel, because each call would
// share access to /dev/stdin. A `readPromise` lock is used to prevent this.
```

Josh-Cena committed:
Update index.md

Co-authored-by: rebloor <git@sherpa.co.nz>

qdirks committed:
Close the fileHandle from within the finally block.

The `try` block is wrapping the calls to `input.read` because those calls can sometimes fail with an error. If they do, then we want to make sure that we still close the open file handle.

qdirks committed:
Use a loop to read all data on larger messages.

Large messages from the extension may not be read all in one go, so you need to loop to get the entire message, and add the bytesRead from each read to an offset value which allows you to continue reading the next chunk of data into your buffer at that offset.

qdirks committed:
Update files/en-us/mozilla/add-ons/webextensions/native_messaging/index.md

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

qdirks committed:
Merge branch 'main' into qdirks-patch-1

qdirks commented:
Oops, I need the loop to be able to get the entire message. But we can do better than the original code by keeping track of the offset into the message buffer where new data needs to be inserted, and by not reading more than the header size by passing that offset to the read method along with the message buffer. I think... I think this is ready now. I'll grant you, the code became longer than I had hoped, but I think this finally solves all of the issues I had.

2 notes which you may want to add about the example:
 - This code won't work on windows, because it accesses a file path (`'/dev/stdin'`) which does not exist on windows.
 - The maximum string size in NodeJS is somewhere around 500 megabytes, at least on my machine. I wasn't able to create strings larger than that. You might want to know that if you're trying to `TextDecoder.decode` the message into a string, and then `JSON.parse` it, since everything the extension sends to your native app is JSON.

 
I may eventually try to re-write this to work on both Linux and Windows using all of the knowledge I gained here, unless somebody beats me to it.

qdirks committed:
Merge branch 'main' into qdirks-patch-1

Josh-Cena commented:
@qdirks I was informed by @guest271314 (who can't participate here) that the `console.error` line in this code won't function in a messaging host, and we should change it to `sendMessage(new TextEncoder().encode(e.message))`. Do you think you could make that update?
